libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -68,6 +76,48 @@ proc waitRepliesOrTimeouts(
 
   return (receivedReplies, failedPeers)
 
+proc dispatchPutVal(
+    kad: KadDHT, peer: PeerId, key: Key, val: seq[byte]
+): Future[void] {.
+    async: (raises: [CancelledError, DialFailedError, ValueError, LPStreamError])
+.} =
+  trace "dialing",
+    me = kad.rtable.selfId, peer = peer, key = key, val = val[0 ..< min(val.len, 4)]
+  let conn = await kad.switch.dial(peer, KadCodec)
+
+  defer:
+    await conn.close()
+
+  let msg = Message(
+    msgType: MessageType.putValue,
+    record: some(Record(key: some(key.getBytes()), value: some(val))),
+  )
+
+  await conn.writeLp(msg.encode().buffer)
+
+  let reply = Message.decode(await conn.readLp(MaxMsgSize)).tryGet()
+
+  if reply.msgType != MessageType.putValue:
+    raise newException(ValueError, "unexpected message type in reply: " & $reply)
+
+proc putVal*(
+    kad: KadDHT, val: seq[byte], replic: int
+): Future[seq[PeerId]] {.async: (raises: [CancelledError]).} =
+  let keyDat: array[IdLength, byte] = sha256.digest(val).data
+  let key = Key(kind: KeyType.Hashed, data: keyDat)
+
```

Comment:
Something to consider is doing here is checking in the local table    if there is an old value already, and it’s not equal to the new one, use the `Select` from the validator (as defined in the spec) to pick the determine whether the old value is prefered to the new one. If the new value is not selected because older is "better", then we can return an error indicating that the value cannot be replaced.